KicadModTree/nodes/specialized/ExposedPad.py

```diff
@@ -0,0 +1,254 @@
+        if type(kwargs.get('paste_layout')) in [int, float]:
+            # when the attribute is a simple number, use it for x and y
+            self.paste_layout = [kwargs.get('paste_layout'), kwargs.get('paste_layout')]
+        else:
+            self.paste_layout = kwargs.get('paste_layout')
+
+        nx = self.paste_layout[0]
+        ny = self.paste_layout[1]
+
+        sx = self.mask_size[0]
+        sy = self.mask_size[1]
+
+        self.paste_size = ExposedPad.__roundToBase(
+                [sx*c/nx, sy*c/ny],
+                kwargs.get('size_round_base', 0.01)
+            )
+
+        dx = (sx - self.paste_size[0]*nx)/(nx+1)
+        dy = (sy - self.paste_size[1]*ny)/(ny+1)
+
+        self.paste_grid = ExposedPad.__roundToBase(
+                [self.paste_size[0]+dx, self.paste_size[1]+dy],
+                kwargs.get('grid_round_base', 0.01)
+            )
+
+    def _createPads(self, **kwargs):
+        pads = []
+        if self.size == self.mask_size:
+            layers_main = ['F.Cu', 'F.Mask']
+        else:
+            layers_main = ['F.Cu']
+            pads.append(Pad(
+                number="", at=self.at, size=self.mask_size,
+                shape=Pad.SHAPE_RECT, type=Pad.TYPE_SMT, layers=['F.Mask']
+            ))
+
+        pads.append(Pad(
+            number=self.number, at=self.at, size=self.size,
+            shape=Pad.SHAPE_RECT, type=Pad.TYPE_SMT, layers=layers_main
+        ))
+
+        cy = -((self.paste_layout[1]-1)*self.paste_grid[1])/2 + self.at.y
+        for i in range(self.paste_layout[1]):
+            pads.append(
+                PadArray(center=[self.at.x, cy],
+                         initial="", increment=0, pincount=self.paste_layout[0],
+                         x_spacing=self.paste_grid[0], size=self.paste_size,
+                         type=Pad.TYPE_SMT, shape=Pad.SHAPE_RECT,
+                         layers=['F.Paste']
+                         ))
+            cy += self.paste_grid[1]
+
+        if self.has_vias:
+            cy = -((self.via_layout[1]-1)*self.via_grid[1])/2 + self.at.y
+            for i in range(self.via_layout[1]):
+                pads.append(
+                    PadArray(center=[self.at.x, cy], initial=self.number,
+                             increment=0, pincount=self.via_layout[0],
+                             x_spacing=self.via_grid[0], size=self.via_size,
+                             type=Pad.TYPE_THT, shape=Pad.SHAPE_CIRCLE,
+                             drill=self.via_drill, layers=['*.Cu']
+                             ))
+                cy += self.via_grid[1]
+
+            if self.add_bottom_pad:
+                pads.append(Pad(
+                    number=self.number, at=self.at, size=self.bottom_size,
+                    shape=Pad.SHAPE_RECT, type=Pad.TYPE_SMT,
+                    layers=self.bottom_pad_Layers
+                ))
+
+        return pads
+
+    @staticmethod
+    def __roundToBase(value, base):
```

Comment:
Duplication of functionality? https://github.com/pointhi/kicad-footprint-generator/blob/master/KicadModTree/Point.py#L64

Reply:
Your function only works for points, not for floats and other values like tubles and arrays.

Reply:
vector is a good idea. Such basic things like round_to should be moved to a utils class where they belong to

Reply:
I changed the name from Point -> Vector and used this new class for all size and grid specifiers in the exposed pad class. This allowed me to remove the roundToBase member from it.